debezium-core/src/main/java/io/debezium/schema/TopicSelector.java

```diff
@@ -82,4 +88,77 @@ public String getHeartbeatTopic() {
     public interface DataCollectionTopicNamer<I extends DataCollectionId> {
         String topicNameFor(I id, String prefix, String delimiter);
     }
+
+    /**
+     * A topic namer that replaces any characters invalid in a topic name with {@code _}.
+     */
+    private static class TopicNameSanitizer<I extends DataCollectionId> implements DataCollectionTopicNamer<I> {
+
+        private static final Logger LOGGER = LoggerFactory.getLogger(TopicNameSanitizer.class);
+
+        private static final String REPLACEMENT_CHAR = "_";
+
+        private final DataCollectionTopicNamer<I> delegate;
+
+        public TopicNameSanitizer(DataCollectionTopicNamer<I> delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public String topicNameFor(I id, String prefix, String delimiter) {
+            String topicName = delegate.topicNameFor(id, prefix, delimiter);
+
+            StringBuilder sanitizedNameBuilder = new StringBuilder(topicName.length());
+            boolean changed = false;
+
+            for(int i = 0; i < topicName.length(); i++) {
+                char c = topicName.charAt(i);
+                if (isValidTopicNameCharacter(c)) {
+                    sanitizedNameBuilder.append(c);
+                }
+                else {
+                    sanitizedNameBuilder.append(REPLACEMENT_CHAR);
+                    changed = true;
+                }
+            }
+
+            if (changed) {
+                String sanitizedName = sanitizedNameBuilder.toString();
```

Comment:
Here's a thing I'm wondering: should we track mapped table names and issue a warning if another table already has been mapped to the same topic name? E.g. two tables named "foo@bar" and "foo$bar" would both be mapped to the topic "foo_bar".

This might not be desirable, e.g. when working with the schema registry. So we could log a warning when detecting this situation, we'd have to keep track of the applied mappings, though. I'm not sure it's worth it.

Note that people can use the table name from the source info block to route such records into distinct topics, if needed. Any thoughts?

Reply:
@gunnarmorling The same problem is with schema registry and schema name adjusting. We should either catch it at both places or write a big red note in the docs. IMHO I think it is enough to document the issue.

Reply:
Ok.